22 crud produits

Lisadev84 opened the pull request:
- Implémentation des fichiers relatifs au Crud produits : classe, repository, service, serviceImpl.
- A ajouter  dans SecurityConfig  API_URLS_WHITELIST pour tester:   "/products/**"

**Modifications apportées :** 
- Ajustement de compose.yml, Dokerfile.dev  back-end(fix conflits et verrouilage de fichiers liés au WSL, ajout d'un container pour les tests )
- ProductDTO 
- Ajout de règles pour l'affichage des produits disponibles
- Ajout des règles pour quantité min et max du stock produit
- Gestion des erreurs
- Tests unitaires de ProductServiceImpl
- Tests unitaires  de ProductController
**Correctif** : 
- Méthode DeleteProduct du service et son test unitaire
- Restructuration de l'entité product : suppression de minQuantité et modification de la propriété MaxQuantity
- Ajustement du service et des tests
- Suppression de la validation OnCreate
- Ajout des tests pour le endpoint status

JeffreyGbeho committed:
initializing entities

JeffreyGbeho committed:
implement register email password

JeffreyGbeho committed:
implement authenticate feature with jwt

JeffreyGbeho committed:
implement authorization with jwt

JeffreyGbeho committed:
implement refresh token feature

JeffreyGbeho committed:
add swagger documentation

JeffreyGbeho committed:
implement google authentication

SayliDev committed:
chore: add react-leaflet and leaflet dependencies

SayliDev committed:
feat(map): initialize Map component with leaflet configuration

SayliDev committed:
feat(map): add markers, user location, and interactive map controls

JeffreyGbeho committed:
implement facebook oauth2 authentication

Lisadev84 committed:
fix compose/dockerfile
create entity Product

Lisadev84 committed:
fix compose/dockerfile
create entity Product

SayliDev committed:
fix(map): resolve type error and refactor file structure

JeffreyGbeho committed:
cleaning up code with solid principle

SayliDev committed:
feat(footer): add basic footer component structure

Lisadev84 committed:
feature:crud product

Lisadev84 committed:
feature:crud product

SayliDev committed:
style(footer): add styling to footer component

SayliDev committed:
style(footer): make footer responsive for mobile devices

Lisadev84 committed:
feature: test ProductController/Postman and fix errors

Lisadev84 committed:
feature: test ProductController/Postman and fix errors

JeffreyGbeho commented on api/src/main/java/com/marketplace/configuration/SecurityConfig.java:
```diff
@@ -21,7 +21,8 @@ public class SecurityConfig {
 
     private static final String[] API_URLS_WHITELIST = {
-            "/test"
+            "/test",
```
Penser à enlever pour la PR. Une bonne alternative est de mettre dans la description de PR les réglages à faire pour pouvoir test

JeffreyGbeho commented on api/src/main/java/com/marketplace/controller/ProductController.java:
Ajouter la documentation swagger sur ton controller avec les annotations @Tag, @Operations, @ResponseAnswers et @Parameters

Lisadev84 replied:
ok, la doc api en français ou en anglais ?

JeffreyGbeho replied:
En anglais vu que c'est une doc qui est destinée pour les dev

JeffreyGbeho commented on api/src/main/java/com/marketplace/controller/ProductController.java:
```diff
@@ -0,0 +1,55 @@
+@RequestMapping("/products")
+public class ProductController {
+
+    @Autowired
```
Avec les nouvelles versions de spring, on utilise plutôt l'injection par constructor qu'avec @Autowired. Ca permet d'avoir un meilleur controle sur l'injection et de pouvoir utiliser final pour bénéficier de l'immutabilité de nos injections.
Pour le faire (vu qu'on utilise Lombok) tu peux ajouter l'annotation @RequiredArgsConstructor et ensuite tu fais ton injection private final ProductService productService;

JeffreyGbeho commented on api/src/main/java/com/marketplace/controller/ProductController.java:
```diff
@@ -0,0 +1,55 @@
+
+    @PostMapping
+    public ResponseEntity<Product> createProduct(@RequestBody Product product) {
+        Product savedProduct = productService.createProduct(product);
```
Une bonne pratique est d'utiliser les logs pour afficher des informations que l'on pourra suivre pour débugger, surtout en prod.
Pour ça t'as une libraire qui s'appelle Slf4j que tu peux ajouter avec l'annotation @slf4j que tu mets au dessus de ta class puis dans tes controller, tu fais quelques choses de ce style
try {
log.info("POST /auth/register - START");
// La logique de ton controller
} finally {
log.info("POST /auth/register - DONE");
}

Le faire dans les controllers ça sera déjà pas mal mais on peut ajouter des logs où on veut

JeffreyGbeho commented on api/src/main/java/com/marketplace/controller/ProductController.java:
```diff
@@ -0,0 +1,55 @@
+    @GetMapping
+    public ResponseEntity<List<Product>> getAllProducts() {
+        List<Product> products = productService.getAllProducts();
+        return products.isEmpty() ? new ResponseEntity<>(HttpStatus.NO_CONTENT) : new ResponseEntity<>(products, HttpStatus.OK);
```
Tu peux faire plus simple. Tu peux juste faire un
return new ResponseEntity<>(products, HttpStatus.OK);
Que la liste soit rempli ou pas, on renvoie un array puis c'est le front qui gère l'affichage du tableau ou du message si c'est vide.
Le code NO_CONTENT est surtout utilisé si ta méthode ne renvoie rien comme les méthodes delete par exemple

JeffreyGbeho commented on api/src/main/java/com/marketplace/controller/ProductController.java:
```diff
@@ -0,0 +1,55 @@
+    private ProductService productService;
+
+    @PostMapping
+    public ResponseEntity<Product> createProduct(@RequestBody Product product) {
```
Pour une question de sécurité et d'optimisation, on évite d'utiliser l'entité pour récupérer les data de la requête. Ca expose des champs que l'on veut pas que l'utilisateur est accès. 
Pour régler le problème, on utilise des DTO (data transfer object) qui seront des objets qui nous d'exposer uniquement les champs que l'on veut et avec la dépendances Validation on va pouvoir utiliser des annotations pour forcer l'utilisateur à envoyer les bonnes infos. 
Si tu veux plus d'infos, n'hésites pas à me dire. C'est un peu long à expliquer en commentaire

JeffreyGbeho commented on api/src/main/java/com/marketplace/controller/ProductController.java:
```diff
@@ -0,0 +1,55 @@
+    @PostMapping
+    public ResponseEntity<Product> createProduct(@RequestBody Product product) {
+        Product savedProduct = productService.createProduct(product);
+        return new ResponseEntity<>(savedProduct, HttpStatus.CREATED);
```
On utilise aussi des DTO pour le renvoi des informations pour une question de sécurité. On veut éviter d'envoyer à l'utilisateur des informations sensibles qui pourrait être contenu dans l'entity (surtout avec les relations entre différentes entity). On pourrait envoyer sans faire exprès un mot de passe. 
Après avoir créer ta class DTO avec les champs que tu veux envoyer. Pour te faciliter la vie tu vas utiliser un mapper de la dépendance Mapstruct (que tu dois créer) qui te permettra de map donc de changer le type de ton objet avec les champs que tu veux. 
Si t'as besoin de plus d'infos, dis moi

JeffreyGbeho commented on api/src/main/java/com/marketplace/controller/ProductController.java:
```diff
@@ -0,0 +1,55 @@
+
+    @PutMapping("/{id}")
+    public ResponseEntity<Product> updateProduct(@PathVariable("id") Long id, @RequestBody Product productDetails) {
+        try {
```
T'as pas besoin de faire un try/catch, les gestions des erreurs sont directement gérer par notre GlobalExceptionHandler. Ca nous permet de centraliser toute la gestion d'erreur à un seul endroit et d'être homogène sur le renvoie du format de réponse

JeffreyGbeho commented on api/src/main/java/com/marketplace/entity/Product.java:
```diff
@@ -0,0 +1,51 @@
+
+
+@Builder
+@Data
```
Une bonne pratique est d'éviter @Data parce que ça génère tout un tas de méthode qui peuvent causer des problèmes avec JPA à la place on préfère utiliser @Getter et @Setter

JeffreyGbeho commented on api/src/main/java/com/marketplace/entity/Product.java:
```diff
@@ -0,0 +1,51 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @NotNull(message = "Le nom du produit ne peut pas être nul")
```
Quand t'auras créer ton DTO tu pourras mettre toutes test constraints dans le DTO

JeffreyGbeho commented on api/src/main/java/com/marketplace/entity/Product.java:
Peut être à voir avec tout le monde, mais je pense qu'il est mieux d'utiliser des noms de variables en anglais vu que tes méthodes sont en anglais. Non ?

JeffreyGbeho commented on api/src/main/java/com/marketplace/service/impl/ProductServiceImpl.java:
```diff
@@ -0,0 +1,60 @@
+
+@Service
+public class ProductServiceImpl implements ProductService {
+    @Autowired
```
Utiliser le Constructor injection à la place du @Autowired

JeffreyGbeho commented on api/src/main/java/com/marketplace/service/impl/ProductServiceImpl.java:
```diff
@@ -0,0 +1,60 @@
+    @Override
+    public Product createProduct(Product product) {
+        if (product.getListeIngredients() != null) {
+            String ingredientsAsString = String.join(", ", product.getListeIngredients().split(", "));
```
Tu peux supprimer cette ligne et utiliser (à la ligne suivante) product.setListeIngredients( product.getListeIngredients());

Parce que là dans ta ligne tu découpes ta string pour ensuite la remettre dans le format initial

JeffreyGbeho commented on api/src/main/java/com/marketplace/service/impl/ProductServiceImpl.java:
```diff
@@ -0,0 +1,60 @@
+                    existingProduct.setSeuilCritique(productDetails.getSeuilCritique());
+                    return productRepository.save(existingProduct);
+                })
+                .orElseThrow(() -> new RuntimeException("Produit non trouvé avec l'ID : " + id));
```
Pour avoir plus de contrôle sur nos exceptions et pouvoir les gérer ensuite dans notre GlobalExceptionHandler, on créer des exceptions qu'on va étendre avec RuntimeException

JeffreyGbeho committed:
clean up and restore application properties configuration

JeffreyGbeho committed:
adding env variables in docker compose file

SayliDev committed:
feat(map): add automatic user geolocation on map load

SayliDev committed:
fix(map): align photo and name in popup, adjust status to align with name

SayliDev committed:
fix(map): reduce map height for mobile version

SayliDev committed:
refactor(map): avoid code duplication in producer status display

SayliDev committed:
fix(footer): remove categories from footer on mobile version

SayliDev committed:
chore: bump version to 0.4.1 for footer fix

SayliDev committed:
fix(map): limit map zoom-out to 50 km radius

SayliDev committed:
chore(version): bump to 0.3.3 for update map

Lisadev84 committed:
modify port

JeffreyGbeho committed:
adding auth controller unit tests

Lisadev84 committed:
add documentation swagger, add injection by constructor

Lisadev84 committed:
add documentation swagger, add injection by constructor

JeffreyGbeho committed:
adding service unit tests

SayliDev committed:
Merge pull request #49 from Pinqo-community/23-producer-map

feat(map): integrate producer map with Leaflet

SayliDev committed:
Merge branch 'develop' into 15-footer

SayliDev committed:
Merge pull request #51 from Pinqo-community/15-footer

feat(footer): implement footer component structure and style

Lisadev84 committed:
add dto et mapper

Lisadev84 committed:
add dto et mapper

JeffreyGbeho committed:
wip

JeffreyGbeho committed:
wip

SayliDev committed:
feat(hero): add Hero Banner component to homepage

SayliDev committed:
chore: update version to 0.5.0

SayliDev committed:
docs: update README with API key setup and frontend technologies

SayliDev committed:
Merge branch 'develop' of https://github.com/Pinqo-community/marketplace-app into develop

SayliDev committed:
feat(ProductCard): Add basic component structure and assets

SayliDev committed:
feat(ProductCard): Add  styling with SCSS modules

SayliDev committed:
feat(ProductCard): Add hover states and transitions

SayliDev committed:
fix(ProductCard): Correct semantic tags

Lisadev84 committed:
fix: ProductController /ProductMapper

Lisadev84 committed:
fix: ProductController /ProductMapper

SayliDev committed:
feat: Add popular products section with product grid layout

SayliDev committed:
chore(release): bump version to 0.6.0

- Add product card component
- Add popular products section in homepage

SayliDev committed:
feat(Category): Add basic component structure and assets

SayliDev committed:
feat(Category): Add styling with SCSS modules

Lisadev84 committed:
fix error mapperProduct - adjust compose.yaml for tests

Lisadev84 committed:
fix error mapperProduct - adjust compose.yaml for tests

Lisadev84 committed:
fix error mapperProduct - adjust compose.yaml for tests

Lisadev84 committed:
fix error mapperProduct - adjust compose.yaml for tests

SayliDev committed:
feat(Category): Add category card in category section

SayliDev committed:
style(Button): Increase button size

SayliDev committed:
style(Product): using relative units for image

SayliDev committed:
chore(release): bump version to 0.6.1

SayliDev committed:
chore(Product): simplify state management and use motion.span for animation

SayliDev committed:
Merge branch '56-product-card' of https://github.com/Pinqo-community/marketplace-app into 56-product-card

JeffreyGbeho committed:
adding swagger schema to dto

JeffreyGbeho committed:
throw Exception when credentials are wrong

JeffreyGbeho committed:
adding TU for checking wrong credentials throw exception

SayliDev committed:
chore(Product): implement API retrieval with mock data

JeffreyGbeho committed:
move expiration token in application properties

Lisadev84 committed:
fix: error configuration Mapper

Lisadev84 committed:
fix: error configuration Mapper

SayliDev committed:
feat: Add Swiper library for implementing sliders

SayliDev committed:
feat(Category): slider implementation

JeffreyGbeho committed:
fix mapper

JeffreyGbeho committed:
fix mapper

SayliDev committed:
feat(Category): added animation to cards and swiper button

SayliDev committed:
fix(ProductImage): use aspect-ratio for consistent square sizing

SayliDev committed:
fix(ProductCard): align note with title on the same line

SayliDev committed:
fix(ProductCard): keep title elements side-by-side on mobile

SayliDev committed:
fix(ProductGrid): switch to grid layout for better spacing

SayliDev committed:
feat(ProductCard): allow user to manually input quantity

Lisadev84 committed:
fix(mapstruct dependencie) : changed order in build.gradle

Lisadev84 committed:
fix(mapstruct dependencie) : changed order in build.gradle

SayliDev committed:
chore(release): bump version to 0.6.2

- Apply requested changes for the PR review

SayliDev committed:
style(hero): update title animation

SayliDev committed:
fix(hero): separate blur only for title side

SayliDev committed:
fix(hero): responsive rule for centered title

SayliDev committed:
chore(release): bump version to 0.5.1

Pops47 reviewed (changes requested):
J'ai un souci quand je lance docker compose depuis cette branche, le front et la bdd se lancent mais pas le back, et j'ai cette erreur dans les logs :  

<img width="1068" alt="Capture d’écran 2024-11-18 à 11 43 15" src="https://github.com/user-attachments/assets/3e8b2bc6-19ec-498f-b21d-24783e4a81d8">

SayliDev committed:
fix(ProductCard): apply aspect-ratio on img tag for Blink
bump version to 0.6.3

Lisadev84 committed:
chore(ProductController): code cleanup; fix(Dockerfile api): chmod +x

Lisadev84 committed:
chore(ProductController): code cleanup; fix(Dockerfile api): chmod +x

SayliDev committed:
fix(Category): box shadow softer on hover

Lisadev84 committed:
feature: error handling added and add an “active” field to the Product entity

Lisadev84 committed:
feature: error handling added and add an “active” field to the Product entity

Lisadev84 committed:
feature:Select products by status and error handling

Lisadev84 committed:
feature:Select products by status and error handling

Lisadev84 committed:
feature:Select products by status and error handling

Lisadev84 committed:
feature:Select products by status and error handling

SayliDev committed:
fix(ProductCard): remove width property, now managed by grid

JeffreyGbeho committed:
fix implementation refresh token

SayliDev committed:
Merge pull request #60 from Pinqo-community/55-hero-banner

feat(hero): add Hero Banner component to homepage

SayliDev committed:
Merge branch '56-product-card' into develop

SayliDev committed:
fix(productCard): adjust title space with note

SayliDev committed:
fix(hero): resolve spacing and alignment issues
- replace margins with a dynamic container
- remove unnecessary rules

SayliDev committed:
fix(sections): resolve spacing and layout inconsistencies

SayliDev committed:
chore: sync 57-category-section with develop (hero and product sections)

Lisadev84 committed:
add property "Id" into ProductDto, create package validation and interfaces, adjust ProductServiceImpl

Lisadev84 committed:
add property "Id" into ProductDto, create package validation and interfaces, adjust ProductServiceImpl

JeffreyGbeho committed:
refactoring TU and cleaning up code

JeffreyGbeho committed:
remove println

Lisadev84 committed:
add ProductControllerTest, adjust dockerfile et compose.yml for tests, adjust ProductDto with annotation

Lisadev84 committed:
add ProductControllerTest, adjust dockerfile et compose.yml for tests, adjust ProductDto with annotation

JeffreyGbeho committed:
Merge pull request #53 from Pinqo-community/17-user-account

17 user account

JeffreyGbeho committed:
Setup redirection route after success oauth authentication

SayliDev committed:
feat(api): integrate categories fetching with RTK Query

- Added `categoriesApi` to manage category-related API requests using RTK Query.
- Configured `categoriesApi.reducerPath` in the Redux store to avoid conflicts.

JeffreyGbeho committed:
setup coverage with jacoco and github actions

JeffreyGbeho committed:
fix coverage verification

JeffreyGbeho committed:
refactor jacoco configuration

SayliDev committed:
feat(Category): make Swiper responsive

SayliDev committed:
fix(Category): resolve TypeScript error with swiper navigation params

SayliDev committed:
chore(release): bump version to 0.7.0

SayliDev committed:
fix(a11y): add accessible label for input and improve form structure

Lisadev84 committed:
add Rule: displays only products with stock above zero
    Rule: Possibility of setting minimum, maximum and stepper quantities

Lisadev84 committed:
add Rule: displays only products with stock above zero
    Rule: Possibility of setting minimum, maximum and stepper quantities

SayliDev committed:
refactor: remove unused class affecting Swiper containers

SayliDev committed:
fix(responsive): set 2 slides per view for mobile display

SayliDev committed:
fix(ui): remove fixed sizes and improve layout for multi-line name

SayliDev committed:
Merge pull request #62 from Pinqo-community/fix/visual-issues

fix: resolve visual issues on homepage sections

SayliDev committed:
feat: add basic component structure and assets

SayliDev committed:
feat: add basic component structure and assets

SayliDev committed:
feat(style): add styling with SCSS modules

SayliDev committed:
feat(style): add styling with SCSS modules

SayliDev committed:
style: organize SCSS properties with Prettier plugin

SayliDev committed:
style: organize SCSS properties with Prettier plugin

SayliDev committed:
feat: add buyer and producer cards to homepage section

SayliDev committed:
feat: add buyer and producer cards to homepage section

SayliDev committed:
feat(animation): add entry animation for BuyerProducerCard

SayliDev committed:
feat(animation): add entry animation for BuyerProducerCard

SayliDev committed:
chore(release): bump version to 0.8.0

SayliDev committed:
chore(release): bump version to 0.8.0

Lisadev84 committed:
add ProductServiceTest and adjust Dockerfile.dev and compose.yaml

Lisadev84 committed:
add ProductServiceTest and adjust Dockerfile.dev and compose.yaml

JeffreyGbeho committed:
remove oauth from the test

JeffreyGbeho committed:
fix configuration jacoco

Lisadev84 committed:
add ProductControllerTest

Lisadev84 committed:
add ProductControllerTest

SayliDev committed:
style(searchBar): Improves UI and use Lucide icon library

SayliDev committed:
style(searchBar): Improves UI and use Lucide icon library

SayliDev committed:
feat(topbar): redesign topbar layout and add new content

SayliDev committed:
feat(topbar): redesign topbar layout and add new content

SayliDev committed:
feat(header): use Lucide icons and add badge for item quantity
- Changing color of the icons
- Breakpoints update for header

SayliDev committed:
feat(header): use Lucide icons and add badge for item quantity
- Changing color of the icons
- Breakpoints update for header

SayliDev committed:
feat(header): add scroll-based visibility toggle

SayliDev committed:
feat(header): add scroll-based visibility toggle

SayliDev committed:
chore: bump frontend version to 0.8.1

SayliDev committed:
chore: bump frontend version to 0.8.1

Lisadev84 committed:
refactoring TU, use WebMvc for ProductControllerTest and cleaning code

Lisadev84 committed:
refactoring TU, use WebMvc for ProductControllerTest and cleaning code

Lisadev84 committed:
refactoring TU, use WebMvc for ProductControllerTest and cleaning code

Lisadev84 committed:
refactoring TU, use WebMvc for ProductControllerTest and cleaning code

Lisadev84 committed:
cleaning code

SayliDev committed:
feat(header): implement compact mode for header

SayliDev committed:
feat(header): implement compact mode for header

JeffreyGbeho commented on api/build.gradle:
```diff
@@ -36,21 +36,24 @@ dependencies {
 	implementation 'org.springframework.boot:spring-boot-starter-oauth2-client'
 	implementation 'org.springframework.boot:spring-boot-starter-security'
 	implementation 'org.springframework.boot:spring-boot-starter-validation'
+	implementation 'jakarta.validation:jakarta.validation-api:3.0.2'
```
Tu peux supprimer cette dépendance.
Jakarta validation est déjà compris dans le spring boot starter validation

JeffreyGbeho commented on api/src/main/java/com/marketplace/controller/ProductController.java:
```diff
@@ -0,0 +1,186 @@
+            List<Product> availableProducts = productService.getAvailableProducts();
+            List<ProductDto> productDtos = productMapper.toDtoList(availableProducts);
+            log.info("GET /products/available - Retrieved {} available products", productDtos.size());
+            if (productDtos.isEmpty()) {
```
Tu peux supprimer ce block de code.
Il vaut mieux renvoyer un json vide avec un code 200 qu'un code 204. C'est plus cohérent et plus simple à gérer pour le front. 
Le code 204 est plus utilisé quand c'est la seule réponse possible sur la route

JeffreyGbeho commented on api/src/test/java/com/marketplace/ApplicationTests.java:
```diff
@@ -1,12 +1,13 @@
-	// @Test
-	// void contextLoads() {
-	// }
+	 @Test
```
Ce test ne fonctionne pas. Je crois que c'est parce qu'on a pas de base de données en test

JeffreyGbeho commented on api/src/test/java/com/marketplace/Service/impl/ProductServiceImplTest.java:
```diff
@@ -0,0 +1,315 @@
+        /**
+         * Custom ArgumentMatcher for Product.
+         */
+        static class ProductMatcher implements ArgumentMatcher<Product> {
```
J'ai pas trop compris l'utilité de cette class. Moyen d'avoir plus d'explications stp ?

JeffreyGbeho commented on api/src/test/java/com/marketplace/controller/ProductControllerTest.java:
```diff
@@ -0,0 +1,229 @@
+                    .andExpect(jsonPath("$.id").value(1L))
+                    .andExpect(jsonPath("$.name").value("Miel de lavande"));
+
+            verify(productService, times(1)).createProduct(any(ProductDto.class));
```
Dans les tests controllers, ce qui est important de tester c'est l'input et l'output. Toutes les méthodes qui sont appelés pour en arriver à l'output ne sont pas testé, ça permet d'avoir des tests moins rigides. 
On utilise verify plutot pour vérifier des appels à des apis externe ou pour des appels méthodes critiques.
Je pense que tu peux supprimer cette ligne et celles des autres tests du controller.

JeffreyGbeho commented on api/src/main/java/com/marketplace/service/impl/ProductServiceImpl.java:
```diff
@@ -0,0 +1,140 @@
+
+        List<Product> products = productRepository.findByActiveAndStockQuantityGreaterThan(true, 0);
+
+        if (products.isEmpty()) {
```
Tu peux enlever ce block. Si ton repository ne trouve aucun produit alors ta list sera déjà empty. Du coup avec le return products tu renverras forcément un empty list

JeffreyGbeho commented on api/src/main/java/com/marketplace/service/impl/ProductServiceImpl.java:
```diff
@@ -0,0 +1,140 @@
+     */
+    @Override
+    public Product getProductById(Long id) {
+        return productRepository.findById(id)
```
Pour gagner en performance et séparer correctement les responsabilités des class, il est préférable de déplacer toute la logique dans ton repository. Par ex : 
@Query("SELECT p FROM Product p WHERE p.id = :id AND p.active = true")
Optional<Product> findByIdAndActive(@Param("id") Long id);

Comme ça dans ton service t'as juste à appeler cette requête et faire un return

JeffreyGbeho commented on api/src/main/java/com/marketplace/service/impl/ProductServiceImpl.java:
```diff
@@ -0,0 +1,140 @@
+
+        List<Product> products = productRepository.findByActive(active);
+
+        if (products.isEmpty()) {
```
Si la liste des produits est vide alors on renvoie une liste vide. N'avoir aucun résultat n'est pas une erreur.
Si on lève une erreur, ça va renvoyer un code d'erreur au front. Le front pourrait traiter les erreurs avec des toasts mais ici on veut pas afficher un toast mais juste un tableau vide avec un message aucun résultat trouvé (par exemple).
NotFound va plus être utilisé si on essaye d'accéder à une ressource spécifique (comme le getById) et que la ressource n'existe pas

JeffreyGbeho commented on api/src/main/java/com/marketplace/service/impl/ProductServiceImpl.java:
```diff
@@ -0,0 +1,140 @@
+     */
+    @Override
+    public void deleteProduct(Long id) {
+        Product product = productRepository.findById(id)
```
Tu peux faire la suppression en une fois avec deleteById ça fait gagner un appel à la db

SayliDev committed:
style: transition time update for cards

SayliDev committed:
Merge branch 'develop' into 57-category-section

SayliDev committed:
Merge pull request #64 from Pinqo-community/57-category-section

feat(Category): Add category section with slider, animations, and API integration

SayliDev committed:
fix: resolve deprecation warning for mixed declarations

SayliDev committed:
Merge branch '58-buyer-producer-section' of https://github.com/Pinqo-community/marketplace-app into 58-buyer-producer-section

SayliDev committed:
chore: update @eslint/plugin-kit to the latest version

Lisadev84 committed:
solving implementation errors, code refactoring

Lisadev84 committed:
solving implementation errors, code refactoring

SayliDev committed:
Merge branch '67-improve-header' into current branch

SayliDev committed:
fix(cards): ensure cards span full width

Lisadev84 committed:
Fix error DeleteProduct method and unit test

Lisadev84 committed:
Fix error DeleteProduct method and unit test

JeffreyGbeho committed:
Merge pull request #66 from Pinqo-community/setup-github-actions

Setup GitHub actions

SayliDev committed:
feat: add buyer/producer cards in homepage

SayliDev committed:
Merge branch 'develop' into 67-improve-header

SayliDev committed:
feat(header): improve header layout

SayliDev committed:
fix: the header width takes the entire screen size

Lisadev84 committed:
Delete .gradle/ideaInitScripts/ijtgtmapper.gradle

Lisadev84 committed:
Delete .gradle/ideaInitScripts/ijtgtmapper.gradle

Lisadev84 committed:
Delete .gradle/ideaInitScripts/ijtgtmapper1.gradle

Lisadev84 committed:
Delete .gradle/ideaInitScripts/ijtgtmapper1.gradle

Lisadev84 committed:
adjust compose.yaml and Map.tsx

Lisadev84 committed:
Rebase develop into crud-product, resolve conflicts

Lisadev84 committed:
Merge remote-tracking branch 'origin/22-crud-produits' into 22-crud-produits

# Conflicts:
#	api/Dockerfile.dev
#	api/build.gradle
#	api/src/main/java/com/marketplace/exception/GlobalExceptionHandler.java
#	api/src/main/resources/application.yml
#	api/src/test/java/com/marketplace/ApplicationTests.java
#	api/src/test/java/com/marketplace/annotation/ControllerWebMvcTest.java
#	api/src/test/java/com/marketplace/controller/ProductControllerTest.java
#	compose.yaml

JeffreyGbeho commented on api/src/main/java/com/marketplace/controller/ProductController.java:
```diff
@@ -0,0 +1,176 @@
+    @Validated(OnCreate.class)
+    @Operation(summary = "Create a product", description = "Add a new product in the marketplace")
+    @ApiResponse(responseCode = "201", description = "Successfully created product")
+    public ResponseEntity<ProductDto> createProduct(@Valid @RequestBody ProductDto productDto) {
```
{
    "name": "tomate",
    "description": "Regardes mon joli panier de tomates",
    "unitPrice": "2",
    "stockQuantity": 10,
    "minQuantity": 5,
    "stepQuantity": 1,
    "criticalLevel": 5,
    "active": true
}

Envoyant ce payload, j'ai un NullPointerException à du maxQuantity. A voir quelle est la meilleure solution pour le gérer. 
Soit rendre le champs obligatoire dans le payload mais est ce que le producteur connait la quantité max à chaque fois ? Sinon gérer le cas dans le service pour éviter cette exception

JeffreyGbeho commented on api/src/main/java/com/marketplace/dto/ProductDto.java:
```diff
@@ -0,0 +1,116 @@
+@Builder
+public class ProductDto {
+
+    @Null(groups = OnCreate.class)
```
Je peux utiliser toutes les routes sans avoir à l'utiliser. Je pense que tu peux le supprimer et du coup ta class OnCreate aussi

JeffreyGbeho commented on api/src/main/java/com/marketplace/repository/ProductRepository.java:
```diff
@@ -0,0 +1,21 @@
+    List<Product> findByActiveAndStockQuantityGreaterThan(boolean active, int stockQuantity);
+
+    @Query("SELECT p FROM Product p WHERE p.id = :id AND p.active = true")
+    Optional<Product> findByIdAndActive(@Param("id") Long id);
```
Commentaire juste à titre d'infos : Tu peux faire findByIdAndActiveIsTrue sans avoir à utiliser @Query ça donnera le même résultat. Mais tu peux laisser comme ça, ça fonctionne aussi

JeffreyGbeho commented on api/src/main/java/com/marketplace/controller/ProductController.java:
```diff
@@ -0,0 +1,176 @@
+
+            })
+    @GetMapping("/status")
+    public ResponseEntity<List<ProductDto>> getProductsByStatus(@RequestParam Boolean active) {
```
Si dans ma requête je n'envoie pas le paramètre active, j'ai une erreur 500. Je pense qu'il faudrait faire en sorte que ça renvoie une erreur 400 en indiquant qu'il manque le paramètre.

JeffreyGbeho commented on api/src/test/java/com/marketplace/controller/ProductControllerTest.java:
```diff
@@ -0,0 +1,196 @@
+@ControllerWebMvcTest (ProductController.class)
+@Import(GlobalExceptionHandler.class)
+@DisplayName("Tests for ProductController")
+class ProductControllerTest {
```
Il te manque les tests pour le endpoint /status

SayliDev reviewed (commented):
L'API répond comme prévu, j'ai testé les différents endpoints (création, récupération, mise à jour, suppression) ainsi que les cas d'erreurs. Mais, lors de la création d'un produit, si le champ `maxQuantity` n'est pas renseigné, aucun message d'erreur (erreur 400) n'est renvoyé alors que le champ devrait être obligatoire ?

Lisadev84 committed:
Fixed NullpointerException errors, restructured MaxQuantity and added endpoint/status tests
